File: lint-rules-rxjava2-lint/src/test/java/com/vanniktech/lintrules/rxjava2/IssueRegistryTest.java
package com.vanniktech.lintrules.rxjava2;

import org.junit.Test;

import static com.vanniktech.lintrules.rxjava2.RxJava2Detector.COMPOSITE_DISPOSABLE_ADD_ALL;
import static com.vanniktech.lintrules.rxjava2.RxJava2Detector.COMPOSITE_DISPOSABLE_DISPOSE;
import static com.vanniktech.lintrules.rxjava2.RxJava2Detector.SUBSCRIBE_MISSING_ERROR_CONSUMER;
import static com.vanniktech.lintrules.rxjava2.RxJava2MethodCheckReturnValueDetector.ISSUE_METHOD_MISSING_CHECK_RETURN_VALUE;
import static com.vanniktech.lintrules.rxjava2.RxJava2MissingCompositeDisposableClearDetector.ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR;
import static org.fest.assertions.api.Assertions.assertThat;

public class IssueRegistryTest {
  @Test public void getIssues() {
    assertThat(new IssueRegistry().getIssues()).containsExactly(
        COMPOSITE_DISPOSABLE_DISPOSE,
        COMPOSITE_DISPOSABLE_ADD_ALL,
        SUBSCRIBE_MISSING_ERROR_CONSUMER,
        ISSUE_METHOD_MISSING_CHECK_RETURN_VALUE,
        ISSUE_MISSING_COMPOSITE_DISPOSABLE_CLEAR
    );
  }
}

File: lint-rules-rxjava2-lint/src/test/java/com/vanniktech/lintrules/rxjava2/IssueRegistryTest.kt
package com.vanniktech.lintrules.rxjava2

import com.android.tools.lint.detector.api.TextFormat.RAW
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class IssueRegistryTest {
  @Test fun everyBriefDescriptionIsASentence() {
    IssueRegistry().issues
      .map { it.getBriefDescription(RAW) }
      .forEach { assertTrue("$it is not a sentence", it.first().isUpperCase() && it.last() == '.' && it == it.trim()) }
  }

  @Test fun everyExplanationConsistsOfSentences() {
    IssueRegistry().issues
      .map { it.getExplanation(RAW) }
      .forEach { assertTrue("$it is not a sentence", it.first().isUpperCase() && it.last() == '.' && it == it.trim()) }
  }

  @Test fun idsStartsWithRxJava2() {
    IssueRegistry().issues
      .map { it.id }
      .forEach { assertTrue("$it is not starting with RxJava2", it.startsWith("RxJava2")) }
  }

  @Test fun idsDoNotHaveDetector() {
    IssueRegistry().issues
      .map { it.id }
      .forEach { assertTrue("$it is containing Detector", !it.contains("Detector")) }
  }

  @Test fun readmeContent() {
    val output = IssueRegistry().issues
      .sortedBy { it.id }
      .joinToString(separator = "\n") { "- **${it.id}** - ${it.getExplanation(RAW)}" }

    assertEquals(
      """
        - **RxJava2DefaultScheduler** - Calling this method will rely on a default scheduler. This is not necessary the best default. Being explicit and taking the overload for passing one is preferred.
        - **RxJava2DisposableAddAllCall** - Instead of using addAll(), add() should be used separately for each Disposable.
        - **RxJava2DisposableDisposeCall** - Instead of using dispose(), clear() should be used. Calling clear will result in a CompositeDisposable that can be used further to add more Disposables. When using dispose() this is not the case.
        - **RxJava2MethodMissingCheckReturnValue** - Methods returning RxJava Reactive Types should be annotated with the @CheckReturnValue annotation. Static analyze tools such as Lint or ErrorProne can detect when the return value of a method is not used. This is usually an indication of a bug. If this is done on purpose (e.g. fire & forget) it should be stated explicitly.
        - **RxJava2MissingCompositeDisposableClear** - A class is using CompositeDisposable and not calling clear(). This can leave operations running and even cause memory leaks. It's best to always call clear() once you're done. e.g. in onDestroy() for Activitys.
        - **RxJava2SchedulersFactoryCall** - Injecting the Schedulers instead of accessing them via the factory methods has the benefit that unit testing is way easier. Instead of overriding them via the Plugin mechanism we can just pass a custom Scheduler.
        - **RxJava2SubscribeMissingOnError** - When calling the subscribe() method an error Consumer should always be used. Otherwise errors might be thrown and may crash the application or get forwarded to the Plugin Error handler.
      """.trimIndent(),
      output
    )
  }
}
